### DISCUSSION TITLE
Can the app.FileSystem access and write files to the pb_hooks directory

### DISCUSSION THREAD

**Discussion by Mashiane** [pocketbase/pocketbase#6177] [Q&A] [answered]
Hi there

I want to automate the process of creating hooks for my app via some template code. So I will extend with javascript using routeAdd.

Is it possible via the app.FileSystem object to access and create files in the pb_hooks directory?

If yes, can I please be advised how?

Thanks in advance.

**Comment by ganigeorgiev** [comment]
The filesystem abstraction is not intended for such purposes (note also that its underlying storage can be S3, not just local filesystem).

It is not clear to me what actually you are trying to do but if you want to access the local filesystem from the `pb_hooks` then you can check the [`$os.*` helpers](https://pocketbase.io/jsvm/modules/_os.html). For more advanced file operations you'll have to switch to Go.

**Comment by Mashiane** [comment]
Thank you so much for the advise... perhaps with the following pseudo code, some clarity can be provided. I have written comments for more green lights.

`
routerAdd("POST", "/api/createhook", (c) => {

    try {

        //get the data from the request

        const data = $apis.requestInfo(c).data

        //get the content for the hook

        const content = data.content

        //get the file name to create

        const fileNme =  data.filename

        //access the hooks directory

        const hooksDir = __hooks

        //build the path to write to

        const hooksPath = hooksDir + '/' + fileNme

        //write content to hooksDir file

        ??

        return c.json(200, { "createhook": "Success" })
    } catch (err) {
        return c.json(400, { "createhook": "Failure", "error": JSON.stringify(err) })
    }
}, $apis.activityLogger($app))
`

The createhook will receieve 2 items, content and filename, then a file will be created in the pb_hooks directory with that filename and content.

I am not even sure of how I can achieve something like this with Go either as I dont have any experience in that. hopefully I will be able to get guidance somewhere.

**Reply by ganigeorgiev** [reply]
Calling [`$os.writeFile(hooksPath, content, 0644)`](https://pocketbase.io/jsvm/functions/_os.writeFile.html) should work.

I still don't understand what you are trying to do and why but couple things:
- PocketBase doesn't have any sandbox features, so you need to be very careful who is allowed to create pb_hooks files because they can do essentially everything (including running custom shell commands or even potentially leaving your application in a broken state).
- Creating or removing hook files while the application is running requires the app process to be restarted so that your change can take effect. There is a file watcher initialized for the `pb_hooks` directory that will attempt to restart the application in case of created/changed/deleted file but it relies on the `execve` command which I think has no equivalent on Windows and we currently print a warning asking users to restart the app manually on Windows.

**Reply by Mashiane** [reply]
Thank you so much, this will do.

**Comment by Mashiane** [comment]
I am close to achieving what I need to do for my app. Thank you so much for your help.

In the pb_hooks directory, I saved a template of the hook that I want to re-create, as a text file. I will read this text file, do a replace of some sections of its contents and then save it back with a new file name.

`//read the contents of the template

        const content = $os.readFile(hooksDir + '/greenapi.txt')

        console.log(content)`

How do I return the file contents as a string? I am getting this

![image](https://github.com/user-attachments/assets/3d4f3589-d479-453f-8ddc-ab250f981345)

I am attaching the text file here for more clarity.

[greenapi.txt](https://github.com/user-attachments/files/18246962/greenapi.txt)

**Reply by ganigeorgiev** [reply, marked as answer]
You can convert it with the `toString(content)` helper.
